File: backend/app/models/content.py
import uuid
from datetime import datetime, timezone
from sqlalchemy import String, Boolean, DateTime, Text, ForeignKey, Integer, JSON
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from app.database import Base


class Subject(Base):
    __tablename__ = "subjects"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    name: Mapped[str] = mapped_column(String(100))
    slug: Mapped[str] = mapped_column(String(100), unique=True, index=True)
    grade_category: Mapped[str] = mapped_column(String(20), index=True)  # primary | junior | senior
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    icon: Mapped[str | None] = mapped_column(String(100), nullable=True)
    color: Mapped[str | None] = mapped_column(String(20), nullable=True)
    order: Mapped[int] = mapped_column(Integer, default=0)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    lessons: Mapped[list["Lesson"]] = relationship(back_populates="subject", cascade="all, delete-orphan")


class Lesson(Base):
    __tablename__ = "lessons"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    subject_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("subjects.id", ondelete="CASCADE"), index=True)
    title: Mapped[str] = mapped_column(String(200))
    slug: Mapped[str] = mapped_column(String(200), index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    content: Mapped[str | None] = mapped_column(Text, nullable=True)   # Rich text / markdown
    order: Mapped[int] = mapped_column(Integer, default=0)
    duration_minutes: Mapped[int] = mapped_column(Integer, default=0)
    is_free_preview: Mapped[bool] = mapped_column(Boolean, default=False)
    is_published: Mapped[bool] = mapped_column(Boolean, default=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc), onupdate=lambda: datetime.now(timezone.utc))

    subject: Mapped["Subject"] = relationship(back_populates="lessons")
    resources: Mapped[list["LessonResource"]] = relationship(back_populates="lesson", cascade="all, delete-orphan")
    progress: Mapped[list["UserProgress"]] = relationship(back_populates="lesson", cascade="all, delete-orphan")  # type: ignore[name-defined]


class LessonResource(Base):
    __tablename__ = "lesson_resources"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    lesson_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("lessons.id", ondelete="CASCADE"))
    resource_type: Mapped[str] = mapped_column(String(20))  # video | pdf | audio | image
    title: Mapped[str] = mapped_column(String(200))
    url: Mapped[str] = mapped_column(Text)
    size_bytes: Mapped[int | None] = mapped_column(Integer, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), default=lambda: datetime.now(timezone.utc))

    lesson: Mapped["Lesson"] = relationship(back_populates="resources")

File: apps/web/src/lib/api/content.ts
'use client';

import { apiRequest } from './client';

export interface Subject {
    id: string;
    name: string;
    slug: string;
    grade_category: string;
    description: string | null;
    icon: string | null;
    color: string | null;
    order: number;
    lesson_count: number;
}

export interface Lesson {
    id: string;
    subject_id: string;
    title: string;
    slug: string;
    description: string | null;
    order: number;
    duration_minutes: number;
    is_free_preview: boolean;
    resources: LessonResource[];
}

export interface LessonResource {
    id: string;
    resource_type: string;
    title: string;
    url: string;
}

export interface LessonDetail extends Lesson {
    content: string | null;
}

export interface SearchResult {
    id: string;
    subject_id: string;
    subject_name: string;
    title: string;
    slug: string;
    description: string | null;
    duration_minutes: number;
    is_free_preview: boolean;
}

export const contentApi = {
    getSubjects: () => apiRequest<Subject[]>('/content/subjects'),
    getLessons: (slug: string) => apiRequest<Lesson[]>(`/content/subjects/${slug}/lessons`),
    getLesson: (lessonId: string) => apiRequest<LessonDetail>(`/content/lessons/${lessonId}`),
    search: (q: string) => apiRequest<SearchResult[]>(`/content/search?q=${encodeURIComponent(q)}`),
};
